### PR TITLE
Logic to switch between RemoteBitrateEstimators

### PR DESCRIPTION
**Pull request by jcchuks**
RemoteBitrateSelector - Logic to switch between RemoteBitrateEstimatorrAbsSendTime and RemoteBitrateEstimatorSingleStream

### CODE REVIEW

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+     *
+     * @return the number of milliseconds until this instance wants a worker
+     * thread to call {@link #run()}
+     */
+    @Override
+    public long getTimeUntilNextRun() {
+        long timeUntilNextRun  =
+                (packetTransformer instanceof RecurringRunnable) ?
+             ((RecurringRunnable) packetTransformer)
+                     .getTimeUntilNextRun() : -1;
```
Comment (bgrozev) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Returning -1 here will result in the executor running this immediately (there are no special semantics for -1, it just executes as long as getTimeUntilNextRun <= 0). I think the easiest solution would be to run it at the same interval as RBESingleStream (500ms). Another option would be to only register with the executor while we're using RBESingleStream, and then de-register when we switch to RBEAbsSendTime.

Comment (jcchuks) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Updated it to return 500Ms. I think it is a better choice. If RBEAbsSendTime should for some reason implement the getTimeUntilNextRun method in the future. Nothing will need to be changed in the Selector.

**Comment by gpolitis** [comment]
This file needs to be reformatted to meet the [Jitsi Coding Convention](https://desktop.jitsi.org/Documentation/CodeConvention). IntelliJ IDEA lets you [reformat source code](https://www.jetbrains.com/help/idea/reformatting-source-code.html) to meet the requirements of your code style, but it looks like you haven't configured it to do so. I have a Jitsi.xml [here](https://gist.github.com/gpolitis/19b75e4814638f8153ab2369a8599549) that you can import in IntelliJIDEA to configure the proper coding style.

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        {
+            ((Runnable)packetTransformer).run();
+        }
+    }
+
+    @Override
+    public void onRttUpdate(long avgRttMs, long maxRttMs) {
+        if(packetTransformer instanceof CallStatsObserver){
+            ((CallStatsObserver)packetTransformer)
+                    .onRttUpdate(avgRttMs,maxRttMs);
```
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Note that the `RemoteBitrateEstimatorAbsSendTime` does not currently implement the `CallStatsObserver`.

**Comment by gpolitis** [comment]
Can we simplify the `RemoteBitrateEstimatorSingleStream` and make it _not_ implement the `TransformEngine` interface?

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+     * The general contract of the method <code>run</code> is that it may
+     * take any action whatsoever.
+     *
+     * @see Thread#run()
+     */
+    @Override
+    public void run() {
+        if (packetTransformer instanceof Runnable)
+        {
+            ((Runnable)packetTransformer).run();
```
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Note that the `RemoteBitrateEstimatorAbsSendTime` does not currently implement the `Runnable`. We should probably _not_ have both the selector and the bitrate estimator implement the `RecurringRunnable` interface.

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+                TransformEngine
+{
+
+    public final static String ENABLE_USE_OF_ABS_SEND_TIME =
+            "org.jitsi.impl.neomedia.rtp.ENABLE_ABS_SEND_TIME";;
+    public static final long kTimeOffsetSwitchThreshold = 30;
+    private static final Logger logger = Logger
+            .getLogger(RemoteBitrateEstimatorSelector.class);
+    private static long defaultTimeUntilNextRunMs = 500;
+    private boolean usingAbsoluteSendTime;
```
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: We usually declare configuration parameters as final.

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    /**
+     * Selects either to use RBE_abs_send_time or RBE_single_stream
+     */
+    private void PickEstimator()
+    {
+        synchronized (critSect)
+        {
+            if (usingAbsoluteSendTime)
```
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: The `synchronized` can be inside the `if` statement, so we don't block if we're not using the ABSRBE.

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+        }
+        //@Todo set default value to false. Allow true for test.
+        enableUseOfAbsSendTime
+                = cfg.getBoolean(ENABLE_USE_OF_ABS_SEND_TIME, false);
+    }
+
+    /**
+     * Selects either to use RBE_abs_send_time or RBE_single_stream
+     */
+    private void PickEstimator()
```
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Capitalization.

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+        /**
+         * RemoteBitrateEstimatorSingleStream is the Default BitrateEstimator
+         */
+        this.packetTransformer
+                = new RemoteBitrateEstimatorSingleStream(observe);
+        this.minBitrateBps = 0;
+        this.packetsSinceAbsoluteSendTime = 0;
+        setAbsSendTimeUsageConfiguration();
+    }
+    private void PickEstimatorFromHeader(RawPacket packet) {
```
Comment (gpolitis) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Capitalization.

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java (hunk 5 above)
Comment (jcchuks) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Does this mean the synchronized construct can be omitted from the `else ` block?

File: src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java (hunk 3 above)
Comment (jcchuks) [inline src/org/jitsi/impl/neomedia/rtp/remotebitrateestimator/RemoteBitrateEstimatorSelector.java]: Currently, only `RemoteBitrateEstimatorSelector ` and `RemoteBitrateEstimatorSingleStream ` implements the `RecurringRunnable ` interface. By your comments, do you mean RecurringRunnable should be removed.

**Pull request closed without merging**